pkg/agent/watch.go

```diff
@@ -201,60 +203,122 @@ func (e vmEvent) Format(state fmt.State, verb rune) {
+		})
 	}
 
-	commonLabelValues := []string{
-		vm.Namespace,
-		vm.Name,
-		vm.Labels[endpointLabel],
+	specMemVal := func(m *int32) *int64 {
+		if m == nil {
+			return nil
+		}
+		memValue := vm.Spec.Guest.MemorySlotSize.Value() * int64(*m)
+		return &memValue
+	}
+	statusMemVal := func(m *resource.Quantity) *int64 {
+		memValue := m.Value()
+		return &memValue
 	}
 
-	// CPU metrics derived from spec
-	specCPU := []kv[vmResourceValueType, *vmapi.MilliCPU]{
-		{key: vmResourceValueMin, value: vm.Spec.Guest.CPUs.Min},
-		{key: vmResourceValueMax, value: vm.Spec.Guest.CPUs.Max},
-		{key: vmResourceValueSpecUse, value: vm.Spec.Guest.CPUs.Use},
+	addMemMetric(vmResourceValueMin, specMemVal(vm.Spec.Guest.MemorySlots.Min))
+	addMemMetric(vmResourceValueMax, specMemVal(vm.Spec.Guest.MemorySlots.Max))
+	addMemMetric(vmResourceValueSpecUse, specMemVal(vm.Spec.Guest.MemorySlots.Use))
+	addMemMetric(vmResourceValueStatusUse, statusMemVal(vm.Status.MemorySize))
+	return metrics
+}
+
+func setVMMetrics(perVMMetrics *PerVMMetrics, vm *vmapi.VirtualMachine, nodeName string) {
+	if vm.Status.Node != nodeName {
+		return
 	}
-	for _, t := range specCPU {
-		if t.value != nil {
-			perVMMetrics.cpu.WithLabelValues(append(commonLabelValues, string(t.key))...).Set(t.value.AsFloat64())
-		}
+
+	cpuMetrics := makeVMCPUMetrics(vm)
+	for _, m := range cpuMetrics {
+		perVMMetrics.cpu.With(m.labels).Set(m.value)
 	}
 
-	if vm.Status.CPUs != nil {
-		perVMMetrics.cpu.WithLabelValues(append(commonLabelValues, string(vmResourceValueStatusUse))...).Set(vm.Status.CPUs.AsFloat64())
+	memMetrics := makeVMMemMetrics(vm)
+	for _, m := range memMetrics {
+		perVMMetrics.memory.With(m.labels).Set(m.value)
 	}
-	// Memory metrics derived from spec
-	specMem := []kv[vmResourceValueType, *int32]{
-		{key: vmResourceValueMin, value: vm.Spec.Guest.MemorySlots.Min},
-		{key: vmResourceValueMax, value: vm.Spec.Guest.MemorySlots.Max},
-		{key: vmResourceValueSpecUse, value: vm.Spec.Guest.MemorySlots.Use},
+}
+
+func updateVMMetrics(perVMMetrics *PerVMMetrics, oldVM, newVM *vmapi.VirtualMachine, nodeName string) {
+	if newVM.Status.Node != nodeName || oldVM.Status.Node != nodeName {
+		// The vm is either has been removed from the nodeName or has been added
+		// there.
+		deleteVMMetrics(perVMMetrics, oldVM, nodeName)
+		setVMMetrics(perVMMetrics, newVM, nodeName)
+		return
 	}
-	for _, t := range specMem {
-		if t.value != nil {
-			memValue := float64(vm.Spec.Guest.MemorySlotSize.Value() * int64(*t.value))
-			perVMMetrics.memory.WithLabelValues(append(commonLabelValues, string(t.key))...).Set(memValue)
+
+	updateMetrics := func(gauge *prometheus.GaugeVec, oldMetrics, newMetrics []vmMetric) {
+		for _, m := range oldMetrics {
+			// this is a linear search, but since we have small number (~10) of
+			// different metrics for each vm, this should be fine.
+			ok := slices.ContainsFunc(newMetrics, func(vm vmMetric) bool {
+				return maps.Equal(m.labels, vm.labels)
+			})
```

Comment:
As written, I'm concerned about how expensive this may be. With #695, it'd result in `2 × 6 × 0.5 × 6 × 4 = 144`[^calculation] average hashmap lookups every time a VM changes.

wdyt about changing `vmMetric.labels` to `[]string` instead of `map[string]string`?
or perhaps instead of `[]vmMetric` using `map[<something>]vmMetric` to make this more efficient?

[^calculation]: 2 types of metrics, 6 "old" metric types, average linear search is only 0.5 the length of the slice (amortized even though order is fixed), 6 "new" metrics, 4 keys in the labels map (and `maps.equal` does lookups proportional to the size of the first map)

Reply:
The reason I used `map[string]string` is that I wanted to keep using `prometheus.Labels` because it makes the code simpler and easier to understand. Following the previous code that used `map[<something>]vmMetric` wasn't straightforward.

About the performance: let's consider the worst case and say we spend 300 ops here. Each op takes 10ns that makes it 300 * 10ns = 3us. Assume we receive 10k VM updates per second (which is an overestimate), then we would spend 30ms on updating metrics. The previous version would have spent 3ms. I don't think the extra 27ms overhead for 10k updates matters here. I'd rather prefer keep the code simpler.

Reply:
That makes sense. Let's merge :rocket: